keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,418 @@
+                fieldPath: spec.nodeName
+      containers:
+      - name: kube-proxy
+        image: kube-proxy-amd64
+        command:
+        - kube-proxy
+        - --config=/etc/config/kube-proxy.yaml
+        - --instance-config=/etc/config/kube-proxy-instance.yaml
+        volumeMounts:
+        - mountPath: /etc/config
+          name: config
+          readOnly: true
+      volumes:
+      - name: config
+        configMap:
+          name: kube-proxy-config-ghkm248tbc
+```
+
+Although the last example is more verbose, it provides the full benefits of
+the ComponentConfig approach, resulting in a consistent surface and versioning
+policy for the component's parameters.
+
+Sometimes, kube-proxy is configured as a static Pod instead of a DaemonSet.
+In that case, instance-specific configuration could be configured by the node
+startup scripts and mounted via a host path volume.
+
+### Kubelet Example
+
+The Kubelet is not typically deployed as a Pod. Instead, startup scripts that
+bootstrap the node write the Kubelet's initial configuration to disk at startup.
+In some deployments, the Kubelet's configuration may later be modified by the
+Dynamic Kubelet Config feature.
+
+Today, the startup script may write a Kubelet command line that looks something
+like `kubelet --config=/etc/config/kubelet.yaml --node-ip=${NODE_IP}`. Under
+this proposal, it would instead write both config files, and a command line like
+`kubelet --config=/etc/config/kubelet.yaml --instance-config=/etc/config/kubelet-instance.yaml`.
+
+When the Dynamic Kubelet Config feature is enabled, the Kubelet only replaces
+the values supplied to `--config` when consuming a remote configuration. Thus,
+the instance-specific parameters would not interfere with sharable parameters
+in this approach.
+
+### Test Plan
+
+ComponentConfig APIs are typically tested similarly to API server APIs, for
+example by testing round-trip conversions. Some tests specifically exercise
+the ability to share a config object, such as the Dynamic Kubelet Config tests.
+
+All existing tests for ComponentConfig Kinds should be extended to include the
+instance-specific Kinds, except where those tests are explicitly testing the
+ability to share a config object across concurrent instances.
+
+### Graduation Criteria
+
+A beta-versioned ComponentConfig API is using the instance-specific object
+approach.
+
+### Upgrade / Downgrade Strategy
+
+Since instance-specific config is currently only exposed on the command-line,
+and no existing ComponentConfig at beta or later maturity levels need to be
+retrofitted to remove instance-specific parameters, immutable upgrades are
+relatively simple: just deploy the new resources with the new configuration
+format (this is how we approached upgrades for ComponentConfig in general).
+Immutable downgrades follow the same approach: deploy the new resources with
+the old version and old config.
+
+In-place upgrades, which are not typically recommended, should not be
```

Comment:
i have not seen data on this, but i think the majority of k8s users are doing in-place either because their tools only work that way or because they don't have the infrastructure to do re-place.
possibly omit `which are not typically recommended`?

Reply:
Yeah, that's fair. I don't have a good idea of what users outside of GKE, which does the immutable node upgrade thing, are doing for upgrades. IIRC the tribal knowledge from back when I was attending SIG-Node was that in-place Kubelet upgrades weren't really supported, and AFAIK it's not what Cluster API does, but neither of those is necessarily a reflection of what people actually do.

Updated the KEP.